piptools/resolver.py

```diff
@@ -219,14 +219,16 @@ def _group_constraints(self, constraints):
             flask~=0.7
 
         """
-        for _, ireqs in full_groupby(constraints, key=key_from_ireq):
+        for _, ireqs in groupby(
+            sorted(
+                constraints,
+                # Within each group, sort by editables first, and within that,
+                # non-constraints before constraints (False sorts before True):
+                key=lambda x: (key_from_ireq(x), not x.editable, x.constraint),
+            ),
+            key=key_from_ireq,
+        ):
             ireqs = list(ireqs)
-            editable_ireq = next((ireq for ireq in ireqs if ireq.editable), None)
-            if editable_ireq:
-                # ignore all the other specs: the editable one is the one that counts
```

Comment:
I couldn't find enough context to know what depended on returning just the first editable, if a reviewer could comment. This would also help determine whether the tie-breaking I added above is necessary or not.